### PR TITLE
Upload/download probe for rucio client #272

### PR DESCRIPTION
**Pull request by nikmagini**
Merge: 94fc77c 6b75e85

### CODE REVIEW

File: tools/probes/common/check_ruciomover [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Authors:
+# - Nicolo Magini, <Nicolo.Magini@cern.ch>, 2017
+
+'''
+Probe to check read/write with all protocols on target RSE with rucio client
+'''
+
+import os
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: we sort our imports alphabetically

python native imports first
then python native imports with from clause

blank line

python external imports
python external from imports

blank line

rucio imports
rucio imports with from clause

File: tools/probes/common/check_ruciomover [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+from rucio.client.replicaclient import ReplicaClient
+
+def upload(rse, protocol, scope, lfn, no_register=False):
+        """
+        Use the rucio upload command to stage out the file.
+        :param rse - destination RSE
+        :param protocol - transfer protocol
+        :param scope - file scope      
+        :param lfn - file lfn
+	:param no_register - skip file registration in Rucio; also disables auto-deletion at lifetime expiration
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: missing one space on vertical alignment ;-)

File: tools/probes/common/check_ruciomover [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+def download(rse, protocol, scope, lfn, dst, pfn=None):
+        """
+        Use the rucio download command to stage in the file.
+        :param rse:   source RSE
+        :param protocol: transfer protocol
+        :param scope: file scope
+        :param lfn:   file lfn
+        :param dst:   local destination directory
+	:param pfn:   file pfn (required when the downloaded file is not registered in Rucio
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: same ;-)

File: tools/probes/common/check_ruciomover [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        """
+	pfnOption=''
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: variable names always_in_camel_case

File: tools/probes/common/check_ruciomover [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        print 'Will use the following command for stageIn:\n%s\n' % cmd
+        s, o = getstatusoutput(cmd)
+        
+        print 'stageIn output:\n%s\n' % o
+
+        if s:
+            raise Exception('stageIn failed -- rucio download did not succeed: %s' % o.replace('\n', ''))
+
+
+OK, WARNING, CRITICAL, UNKNOWN = 0, 1, 2, 3
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: should be moved into a python __main__

File: tools/probes/common/check_ruciomover [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+parser=argparse.ArgumentParser()
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: hmmm pylint should have given you "missing spaces around operator" warning here. did you run a recent version of pylint against the file?

File: tools/probes/common/check_ruciomover [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+	sys.exit(UNKNOWN)
+
+print 'Testing RSE:\n%s\n' % rse
+
+scope='tests'
+
+timestamp='%d' % time.time()
+fileuuid=str(uuid.uuid1())
+
+reportfile='%s/nagios-check_ruciomover.%s.%s.csv' % (os.getcwd(), rse, timestamp)
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: @cserf is this ok for nagios?

File: tools/probes/common/check_ruciomover [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+if not args.testmode:
+	try:
+		proxy = config_get('nagios', 'proxy')
+		os.environ["X509_USER_PROXY"] = proxy
+	except Exception as e:
+		print "Failed to get proxy from rucio.cfg"
+		sys.exit(UNKNOWN)
+
+try:
+        rse_settings = rsemanager.get_rse_info(rse)
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: wrong indentation?

File: tools/probes/common/check_ruciomover [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+		print 'Skipping sitemover tests on tape RSE'
+		sys.exit(UNKNOWN)
+	
+	rse_protocols=rse_settings['protocols']
+
+	# Check if the DID selected for the download test is present on the target RSE
+	skip_download=False
+	if not args.download_uploaded:
+		repcli = ReplicaClient()
+		reps = [r for r in repcli.list_replicas([{'scope': args.download.split(':')[0], 'name': args.download.split(':')[1]}])]
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: split long lines on commas

File: tools/probes/common/check_ruciomover [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+		if len(reps)==0 or rse not in reps[0]['rses'].keys():
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: if not reps instead of len(reps)==0

File: tools/probes/common/check_ruciomover [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+	    # Run tests only if the RSE supports read/write for this protocol.
+	    # Need to check for wan since the nagios probe is running remotely. 	
+            if not (prot['domains']['wan']['read'] and prot['domains']['wan']['write']):
+                continue
+
+	    print 'Testing protocol:\n%s\n' % prot['scheme']
+
+	    report[prot['scheme']]={'upload':'UNKNOWN', 'download':'UNKNOWN'}
+
+            filename='%s.nagios-check_ruciomover.%s.%s.%s.%s' % (scope, rse, prot['scheme'], timestamp, fileuuid)
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: indentation?

File: tools/probes/common/check_ruciomover [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+		    upload(rse, prot['scheme'], scope, filename, no_register=args.no_register)
+		    report[prot['scheme']]['upload']='OK'
+	    except Exception as e:
+		    print '\n%s\n' % e
+		    report[prot['scheme']]['upload']='CRITICAL'
+		    if args.download_uploaded:
+			    os.unlink(filename)
+			    continue
+	    
+            dst='download/%s.%s.%s' % (rse, prot['scheme'], timestamp)
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: indentation?

File: tools/probes/common/check_ruciomover [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+	  
+	    pfntodownload=None
```
Comment (mlassnig) [inline tools/probes/common/check_ruciomover]: camel_case

**Review by mlassnig — COMMENTED** [review]
So the probe does the right thing, that's good, thanks! Just a few minor comments regarding code style, nothing major. Also, can you switch to 4-spaces, instead of what I see as 8-tab?

PS: please rename the pull-request so we know what it is about ;-) also, this is a general rucio-clients test probe, not specific to ruciomover

**Comment by mlassnig** [comment]
Just in case you haven't seen it, you can look at the Travis errors in detail. This is what's left:


tools/probes/common/check_ruciomover:31:1: E302 expected 2 blank lines, found 1

tools/probes/common/check_ruciomover:54:1: E302 expected 2 blank lines, found 1

tools/probes/common/check_ruciomover:84:1: E302 expected 2 blank lines, found 1

tools/probes/common/check_ruciomover:85:9: F841 local variable 'WARNING' is assigned to but never used

tools/probes/common/check_ruciomover:169:47: E231 missing whitespace after ':'

tools/probes/common/check_ruciomover:169:69: E231 missing whitespace after ':'

tools/probes/common/check_ruciomover:179:66: E231 missing whitespace after ':'

tools/probes/common/check_ruciomover:179:80: E231 missing whitespace after ':'

tools/probes/common/check_ruciomover:219:13: E303 too many blank lines (2)

tools/probes/common/check_ruciomover:234:13: E303 too many blank lines (2)

tools/probes/common/check_ruciomover:261:1: E305 expected 2 blank lines after class or function definition, found 1

**Comment by mlassnig** [comment]
Down to one problem :-)


tools/probes/common/check_ruciomover:298:1: E305 expected 2 blank lines after class or function definition, found 1

**Comment by vingar** [comment]
you can install locally the pre-commit hook to get such errors before
With  ./tools/configure_git.sh 
or 
cp tools/pre-commit .git/hooks/pre-commit
chmod +x .git/hooks/pre-commit

**Comment by nikmagini** [comment]
Thanks for the notification, fixed, somehow flake8 didn't report it for me... Anyway I'm now also setting up the development env properly

**Comment by mlassnig** [comment]
can you do

git fetch --all
git rebase upstream/master

and then try again with pylint -E on your local install?

**Comment by nikmagini** [comment]
Done, and also took the opportunity to squash my commits into a single one. I didn't get any errors from pylint -E, let's see if Travis agrees.

**Review by mlassnig — APPROVED** [review]

**Pull request merged**